feat(avm): tag checking, raising errors and stop execution

jeanmon opened the pull request:
Resolves #9745

Commit:
9745: Remove magic constant in AvmExecutionTests.powerWithMulOpcodes

Commit:
9745: Raise op_err flag when tag_err is raised

Commit:
9745: Handle bitwise opcodes

Commit:
9745: additional opcodes - accrued substates

Commit:
9745: external call opcode

Commit:
9745: tag checking on remaining opcodes

Commit:
9745: stop execution when an error occurs

Commit:
9745: resolves rebase to master conflicts

Commit:
9745: fix tests

dbanks12 reviewed (approved):
Nice work! I commented with some questions and notes.

In general, I think this is a big step closer to what we want. I do wonder if we could avoid having this error return type and instead let opcodes raise a "reverted" field on the TraceBuilder class. This is similar to how in the simulator opcodes raise `context.machineState.reverted`. And then at the bottom of the execution loop, we check `trace_builder.reverted` before continuing?

I don't feel strongly that you need to do it this way, especially not as a prerequisite to merge. BUT it would be a much smaller changeset I think. 

I'm also surprised that you can change the tag types in a handful of places without changing brillig generation.

jeanmon commented on yarn-project/simulator/src/avm/opcodes/ec_add.ts:
```diff
@@ -80,16 +83,11 @@ export class EcAdd extends Instruction {
-      dest = p2;
-    } else if (p2IsInfinite) {
-      dest = p1;
-    }
+
```
It was a leftover of a previous version. This is exactly a subset of the "if ... else if ... else ..." clauses on the previous lines.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/tests/cast.test.cpp:
```diff
@@ -184,6 +184,7 @@ TEST_F(AvmCastTests, truncationFFToU16ModMinus1)
     trace_builder = AvmTraceBuilder(public_inputs, {}, 0, calldata)
                         .set_full_precomputed_tables(false)
                         .set_range_check_required(false);
+    trace_builder.op_set(0, 0, 0, AvmMemoryTag::U32);
     trace_builder.op_set(0, 1, 1, AvmMemoryTag::U32);
     trace_builder.op_calldata_copy(0, 0, 1, 0);
     trace_builder.op_cast(0, 0, 1, AvmMemoryTag::U16);
```
Because now, we enforce that some offsets passed to CALLDATACOPY have tag U32. It was working before because memory is per default initialized with zero of type FF.

jeanmon commented on barretenberg/cpp/pil/avm/main.pil:
```diff
@@ -507,7 +515,7 @@ namespace main(256);
     pol commit sel_slice_gadget; // Selector to activate a slice gadget operation in the gadget (#[PERM_MAIN_SLICE]).
 
     // Activate only if tag_err is disabled or retsize (ib) is non-zero
```
Good catch.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp:
```diff
@@ -1695,7 +1748,7 @@ void AvmTraceBuilder::op_returndata_copy(uint8_t indirect,
 // Helper for "gas left" related opcodes
 void AvmTraceBuilder::execute_gasleft(EnvironmentVariable var, uint8_t indirect, uint32_t dst_offset)
 {
-    assert(var == EnvironmentVariable::L2GASLEFT || var == EnvironmentVariable::DAGASLEFT);
+    ASSERT(var == EnvironmentVariable::L2GASLEFT || var == EnvironmentVariable::DAGASLEFT);
```
Related to assert vs. ASSERT: It does not change anything AFAIK, but the latter was usually the adopted style throughout our code base.

It is really an assertion and not error handling because this function is called only from the externally exposed routines: op_l2gasleft() and op_dagasleft

In other words, this is a coding mistake.

Related to `op_get_env_var`, it is guarded by the first if. the "default" case in the switch is to make the compiler happy but we have the guarantee that it cannot happen. I will add a comment about this.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp:
```diff
@@ -2126,20 +2185,15 @@ Row AvmTraceBuilder::create_kernel_output_opcode_with_metadata(uint8_t indirect,
 
-    // TODO: resolved_address should be used
-    auto [resolved_data, resolved_address, resolved_metadata] =
-        Addressing<3>::fromWire(indirect, call_ptr)
-            .resolve({ data_offset, address_offset, metadata_offset }, mem_trace_builder);
-
     auto read_a = constrained_read_from_memory(
-        call_ptr, clk, resolved_data, AvmMemoryTag::FF, AvmMemoryTag::U1, IntermRegister::IA);
+        call_ptr, clk, data_offset, AvmMemoryTag::FF, AvmMemoryTag::U1, IntermRegister::IA);
 
     auto write_b = constrained_write_to_memory(
-        call_ptr, clk, resolved_metadata, exists, AvmMemoryTag::FF, AvmMemoryTag::U1, IntermRegister::IB);
+        call_ptr, clk, metadata_offset, exists, AvmMemoryTag::FF, AvmMemoryTag::U1, IntermRegister::IB);
     bool tag_match = read_a.tag_match && write_b.tag_match;
```
Because we moved it to the caller. This helper is only called by a single caller (op_nullifier_exists).

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp:
```diff
@@ -3212,39 +3449,42 @@ void AvmTraceBuilder::op_sha256_compression(uint8_t indirect,
+    bool op_valid = tag_match;
+    for (uint32_t i = 0; i < KECCAKF1600_INPUT_SIZE; i++) {
+        op_valid = op_valid && unconstrained_get_memory_tag(resolved_input_offset + i) == AvmMemoryTag::U64;
+    }
 
     // Constrain gas cost
     gas_trace_builder.constrain_gas(clk, OpCode::KECCAKF1600);
 
     main_trace.push_back(Row{
         .main_clk = clk,
-        .main_ia = input_read.val,  // First element of input
-        .main_ic = output_read.val, // First element of output
+        .main_ia = input_read.val, // First element of input
         .main_ind_addr_a = FF(input_read.indirect_address),
-        .main_ind_addr_c = FF(output_read.indirect_address),
         .main_internal_return_ptr = FF(internal_return_ptr),
         .main_mem_addr_a = FF(input_read.direct_address),
-        .main_mem_addr_c = FF(output_read.direct_address),
+        .main_op_err = FF(static_cast<uint32_t>(!op_valid)),
         .main_pc = FF(pc),
         .main_r_in_tag = FF(static_cast<uint32_t>(AvmMemoryTag::U64)),
         .main_sel_mem_op_a = FF(1),
-        .main_sel_mem_op_c = FF(1),
         .main_sel_op_keccak = FF(1),
         .main_sel_resolve_ind_addr_a = FF(static_cast<uint32_t>(input_read.is_indirect)),
-        .main_sel_resolve_ind_addr_c = FF(static_cast<uint32_t>(output_read.is_indirect)),
         .main_tag_err = FF(static_cast<uint32_t>(!tag_match)),
```
The output offset is only required to write the result. There is no reason to read anything here. The simulator code does not do it and I noticed this discrepancy while working on this task. As these gadgets are heavily stubbed/underconstrained, it was never noticed I guess.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp:
```diff
@@ -3844,12 +4142,16 @@ std::vector<Row> AvmTraceBuilder::finalize()
     for (size_t i = 0; i < new_trace_size; i++) {
         auto& r = main_trace.at(i);
 
+        if (r.main_tag_err == FF(1)) {
+            r.main_op_err = FF(1); // Consolidation of errors into main_op_err
+        }
+
         if ((r.main_sel_op_add == FF(1) || r.main_sel_op_sub == FF(1) || r.main_sel_op_mul == FF(1) ||
```
When we constrain yes. The tag_err is very specific and it is easier to reason about to isolate from other potential errors. I think it would be diificult to use a single error flag. In addition, we have specific constraints related to tag_err that I had to use op_err for the non-constrained tag errors, otherwise we would have wrong constraints.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/tests/execution.test.cpp:
```diff
@@ -1743,8 +1756,9 @@ TEST_F(AvmExecutionTests, kernelOutputEmitOpcodes)
         std::ranges::find_if(trace.begin(), trace.end(), [](Row r) { return r.main_sel_op_emit_unencrypted_log == 1; });
     ASSERT_TRUE(emit_log_row != trace.end());
 
-    // Trust me bro for now, this is the truncated sha output
-    FF expected_hash = FF(std::string("0x00b5c135991581f3049df936e35ef23af34bb04a4775426481d944d35a618e9d"));
+    // Warning: The following hash depends on the contract instance address, i.e., any modifications of the bytecode
+    // above will affect this hash.
+    FF expected_hash = FF(std::string("0x00866d9ba8bbca359589c02f03188a2b8368c3e1284a714b9e65b644063c86ab"));
```
The following PR removes the hardcoded string:  #9813

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp:
```diff
@@ -2459,20 +2533,41 @@ void AvmTraceBuilder::op_emit_note_hash(uint8_t indirect, uint32_t note_hash_off
-                                          uint32_t address_offset,
-                                          uint32_t dest_offset)
+AvmError AvmTraceBuilder::op_nullifier_exists(uint8_t indirect,
+                                              uint32_t nullifier_offset,
+                                              uint32_t address_offset,
+                                              uint32_t dest_offset)
 {
     auto const clk = static_cast<uint32_t>(main_trace.size()) + 1;
 
-    Row row = create_kernel_output_opcode_with_set_metadata_output_from_hint(
-        indirect, clk, nullifier_offset, address_offset, dest_offset);
-    kernel_trace_builder.op_nullifier_exists(
-        clk, side_effect_counter, row.main_ia, /*safe*/ static_cast<uint32_t>(row.main_ib));
-    row.main_sel_op_nullifier_exists = FF(1);
+    // TODO: resolved_address should be used
+    auto [resolved_nullifier_offset, resolved_address, resolved_dest] =
+        Addressing<3>::fromWire(indirect, call_ptr)
+            .resolve({ nullifier_offset, address_offset, dest_offset }, mem_trace_builder);
+
+    bool op_valid = unconstrained_get_memory_tag(resolved_address) == AvmMemoryTag::FF;
+
+    Row row;
+
+    if (op_valid) {
+        row = create_kernel_output_opcode_with_set_metadata_output_from_hint(
+            clk, resolved_nullifier_offset, resolved_address, resolved_dest);
+        kernel_trace_builder.op_nullifier_exists(
+            clk, side_effect_counter, row.main_ia, /*safe*/ static_cast<uint32_t>(row.main_ib));
+        row.main_sel_op_nullifier_exists = FF(1);
+        op_valid = op_valid && row.main_tag_err == FF(0);
+    } else {
+        row = Row{
+            .main_clk = clk,
+            .main_internal_return_ptr = internal_return_ptr,
+            .main_op_err = FF(1),
+            .main_pc = pc,
+            .main_sel_op_nullifier_exists = FF(1),
+        };
+    }
 
```
Yep I removed it as it is no longer valid.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp:
```diff
@@ -2974,6 +3128,7 @@ std::vector<FF> AvmTraceBuilder::op_revert(uint8_t indirect, uint32_t ret_offset
         slice_trace_builder.create_return_slice(returndata, clk, call_ptr, resolved_ret_offset, ret_size);
     }
 
+    // TODO: fix and set sel_op_revert
     main_trace.push_back(Row{
         .main_clk = clk,
         .main_call_ptr = call_ptr,
```
Not completely trivial. If I just change the op selector, then the return slice is not properly constrained. It is also not a huge task but I would keep it separately.
I opened #9896 for this.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp:
```diff
@@ -2661,46 +2771,71 @@ void AvmTraceBuilder::op_emit_unencrypted_log(uint8_t indirect,
-    for (uint32_t i = 0; i < log_size; i++) {
-        FF log_value = unconstrained_read_from_memory(direct_field_addr + i);
-        std::vector<uint8_t> log_value_byte = log_value.to_buffer();
+    bool op_valid = unconstrained_get_memory_tag(resolved_log_offset) == AvmMemoryTag::FF &&
```
The changes consist of making the tag checks:
- resolved_log_size_offset has tag U32
- On range [direct_field_addr,  direct_field_addr + log_size), check tag FF

and guarding some code flow if there is a failure.

jeanmon commented:
> Nice work! I commented with some questions and notes.
> 
> In general, I think this is a big step closer to what we want. I do wonder if we could avoid having this error return type and instead let opcodes raise a "reverted" field on the TraceBuilder class. This is similar to how in the simulator opcodes raise `context.machineState.reverted`. And then at the bottom of the execution loop, we check `trace_builder.reverted` before continuing?
> 
> I don't feel strongly that you need to do it this way, especially not as a prerequisite to merge. BUT it would be a much smaller changeset I think.
> 
> I'm also surprised that you can change the tag types in a handful of places without changing brillig generation.

@dbanks12 True, setting the error/revert in trace_builder and having execution loop checking it at the end of the loop would have been less tedious change. The current change which sets the error at each opcode function call could have been avoided. As we plan a big refactoring in any case, I would not revert this now (especially you mentioned you can live with it.)
.cc @fcarreiro for awareness.

Commit:
9745: Addressing review feedback

Commit:
9745: refactoring to introduce check_tag routines

Commit:
9745: conflict resolution after rebase on master

Commit:
9745: Do not restrict Field tag in return opcode
